### PR TITLE
Add Unit Tests

### PR DESCRIPTION
**Pull request by Fl1s**
Add a bunch of tests for directories:
/config/
/core/http/
/core/https/ -> I had serious troubles with this one 
/exception/
/utils/

result

I'm having some difficulties with HttpServerListenerThreadTest and HttpsServerListenerThreadTest in particular. 
Namely:

HttpServerListenerThreadTest -
`java.net.BindException: Address already in use`
test1
I forgot how to release the port after a test, and therefore need to fix the @AfterEach block to do this.

HttpsServerListenerThreadTest - I fundamentally did not make any changes to the core logic, writing only code for tests. Therefore, I did not create getters and setters methods. I also could not test the private method createSSLServerSocket() because of its access modifier. And lastly, I think we need method to close serverListenerThread.

### CODE REVIEW

**Comment by Guhapriya01** [comment]
@Fl1s, 

I will refactor the `createSSLServerSocket()` method and add getters to the server listener threads so that you can test them. Also, we can change the `tearDown` method to help fix the `BindException` you’re seeing.

Could you please try the updated `tearDown` method below and see if it works?

```java
@AfterEach
public void tearDown() throws IOException {
    if (serverListener.isAlive()) {
        serverListener.interrupt(); // Stop the thread if it's running
    }
    if (serverSocket != null && !serverSocket.isClosed()) {
        serverSocket.close(); 
    }
    if (clientSocket != null && !clientSocket.isClosed()) {
        clientSocket.close(); 
    }
}
```

Let me know how it goes!

**Comment by Fl1s** [comment]
Unfortunately, the address is still in use:
tests

**Comment by Guhapriya01** [comment]
> Gm @Guhapriya01! Unfortunately, the address is still in use: tests


@Fl1s, I see that you’re still encountering the `BindException`. Sometimes, adding a short delay before the next test starts can help resolve this issue.

Could you try adding this line at the end of the `tearDown` method?

```java
Thread.sleep(100); 
```

Let me know if this helps!

**Comment by Fl1s** [comment]
Nah, that didn't help, but I figured out where the bug was hiding:
`serverListener = new HttpServerListenerThread(0, "/webroot")`
It turned out that when I specified port 8080, it was full. As soon as I changed it to any free port(0), all tests passed!
tests

**Comment by Guhapriya01** [comment]
@Fl1s, I’ve changed the access modifier of the createSSLServerSocket method and added getters. Could you make the necessary changes to the HttpsServerListenerThreadTest accordingly?

**Comment by Fl1s** [comment]
> @Fl1s, I’ve changed the access modifier of the createSSLServerSocket method and added getters. Could you make the necessary changes to the HttpsServerListenerThreadTest accordingly?

Done 😼

**Comment by Guhapriya01** [comment]
> Done 😼

Since there were some changes since the last code coverage report, could you please share the updated one?

**Comment by Fl1s** [comment]
> > Done 😼
>  
> Since there were some changes since the last code coverage report, could you please share the updated one?

Check my last commit `Add more tests + optimized it`

**Comment by Guhapriya01** [comment]
> Check my last commit `Add more tests + optimized it`

@Fl1s, everything looks good! I would just like to see the updated code coverage. Could you please share it?

**Comment by Fl1s** [comment]
Oh, right
tests
Just In HttpsServerListenerThreadTest, you will need to insert SSLConfiguration

**Comment by Guhapriya01** [comment]
@Fl1s,

I noticed an issue while running the tests for the `HttpsServerListenerThread` class.

error message

Are you aware of this issue? It seems like we might need to add `@AfterEach` to deregister the static mock. Am I missing anything in my setup, or could you address this error in your code?

Thanks!

**Comment by Fl1s** [comment]
Try this one:
```java
@AfterEach
public void tearDown() {
    if (mockedSSLFactory != null) {
        mockedSSLFactory.close();
    }
}
```
If everything works fine, I'll upload it to a new commit

**Comment by Guhapriya01** [comment]
@Fl1s,

Yes! This worked:

```java
@AfterEach
public void tearDown() {
    mockedConfigManager.close();
    mockedSSLFactory.close();  
}
```

Also, please change the `IOException` to `RuntimeException` in the `testSSLServerSocketCreationFailure` method. Since `SSLServerSocketFactory.getDefault()` does not throw checked exceptions, we should mock it to throw an unchecked exception (like `RuntimeException`).

Here’s the error I encountered:

```
org.mockito.exceptions.base.MockitoException: 
Checked exception is invalid for this method!
Invalid: java.io.IOException: Failed to create SSL Server Socket
	at com.httpserver.core.https.HttpsServerListenerThreadTest.testSSLServerSocketCreationFailure(HttpsServerSocketCreationTest.java:90)
```

**Comment by Fl1s** [comment]
That's kind of it now...Whew

File: src/test/java/com/httpserver/core/https/HttpsServerListenerThreadTest.java [hunk 1]
```diff
@@ -24,111 +25,89 @@ public class HttpsServerListenerThreadTest {
...
-
-            IOException exception = assertThrows(IOException.class, serverListenerThread::createSSLServerSocket);
-
-            assertAll("SSL Server Socket Creation Failure",
-                    () -> assertEquals("Failed to create SSL Server Socket", exception.getMessage(), "Exception message should match"),
-                    () -> assertTrue(exception != null, "Exception should be of type IOException")
-            );
-        }*/
+    public void testSSLServerSocketCreationFailure() throws RuntimeException {
+        mockedSSLFactory.when(SSLServerSocketFactory::getDefault).thenThrow(new IOException("Failed to create SSL Server Socket"));
```
Comment (Guhapriya01) [inline src/test/java/com/httpserver/core/https/HttpsServerListenerThreadTest.java:85]: Change IOException to RuntimeException

**Review by Guhapriya01 — APPROVED** [review]

**Comment by Guhapriya01** [comment]
@Fl1s,

I really appreciate your contribution to this project. Great work! Your efforts in enhancing our test coverage are invaluable, and these updates will help ensure the reliability of our project.

Thanks again for your contribution! ❤️

**Comment by Fl1s** [comment]
🥳

**Pull request merged**